src/trace/span-inferrer.ts

```diff
@@ -0,0 +1,401 @@
+import {
+  Context,
+  DynamoDBStreamEvent,
+  EventBridgeEvent,
+  KinesisStreamEvent,
+  S3CreateEvent,
+  SNSEvent,
+  SNSMessage,
+  SQSEvent,
+} from "aws-lambda";
+import { SpanContext, SpanOptions, TracerWrapper } from "./tracer-wrapper";
+import { eventSources, parseEventSource } from "./trigger";
+import { SpanWrapper } from "./span-wrapper";
+export class SpanInferrer {
+  traceWrapper: TracerWrapper;
+  constructor(traceWrapper: TracerWrapper) {
+    this.traceWrapper = traceWrapper;
+  }
+
+  public createInferredSpan(event: any, context: Context | undefined, parentSpanContext: SpanContext | undefined): any {
+    const eventSource = parseEventSource(event);
+    if (eventSource === eventSources.apiGateway) {
+      return this.createInferredSpanForApiGateway(event, context, parentSpanContext);
+    }
+    if (eventSource === eventSources.sns) {
+      return this.createInferredSpanForSns(event, context, parentSpanContext);
+    }
+    if (eventSource === eventSources.dynamoDB) {
+      return this.createInferredSpanForDynamoDBStreamEvent(event, context, parentSpanContext);
+    }
+    if (eventSource === eventSources.sqs) {
+      return this.createInferredSpanForSqs(event, context, parentSpanContext);
+    }
+    if (eventSource === eventSources.kinesis) {
+      return this.createInferredSpanForKinesis(event, context, parentSpanContext);
+    }
+    if (eventSource === eventSources.s3) {
+      return this.createInferredSpanForS3(event, context, parentSpanContext);
+    }
+    if (eventSource === eventSources.eventBridge) {
+      return this.createInferredSpanForEventBridge(event, context, parentSpanContext);
+    }
+  }
+
+  isApiGatewayAsync(event: any): boolean {
+    return (
+      event.headers && event.headers["X-Amz-Invocation-Type"] && event.headers["X-Amz-Invocation-Type"] === "Event"
+    );
+  }
+
+  createInferredSpanForApiGateway(
+    event: any,
+    context: Context | undefined,
+    parentSpanContext: SpanContext | undefined,
+  ): SpanWrapper {
+    const options: SpanOptions = {};
+    const domain = event.requestContext.domainName;
+    const path = event.rawPath || event.requestContext.routeKey;
+    let method;
+    if (event.requestContext.httpMethod) {
+      method = event.requestContext.httpMethod;
+    } else if (event.requestContext.http) {
+      method = event.requestContext.http.method;
+    }
+    const resourceName = [domain, path].join(" ");
+    options.tags = {
+      operation_name: "aws.apigateway",
+      "http.url": domain + path,
+      endpoint: path,
+      resource_names: resourceName,
+      request_id: context?.awsRequestId,
+      "span.type": "http",
+      "resource.name": resourceName,
+      "service.name": domain,
+      api_id: event.requestContext.apiId,
```

Comment:
Can we rename this to `apiid`? This is done since the metric dimension is `apiid` and we want to have an attribute that's the same name as the metric tag. This ensure that the attribute is there when pivoting from a metrics to a trace query.
```suggestion
      apiid: event.requestContext.apiId,
```